src/deploy/functions/deploymentPlanner.ts

```diff
@@ -0,0 +1,175 @@
+import * as deploymentTool from "../../deploymentTool";
+import { functionMatchesAnyGroup, getTopicName } from "../../functionsDeployHelper";
+
+export interface CloudFunctionTrigger {
+  name: string;
+  sourceUploadUrl?: string;
+  labels: { [key: string]: string };
+  environmentVariables: { [key: string]: string };
+  entryPoint: string;
+  runtime?: string;
+  vpcConnector?: string;
+  vpcConnectorEgressSettings?: string;
+  ingressSettings?: string;
+  availableMemoryMb?: number;
+  timeout?: number;
+  maxInstances?: number;
+  serviceAccountEmail?: string;
+  httpsTrigger?: any;
+  eventTrigger?: any;
+  failurePolicy?: {};
+  schedule?: object;
+  timeZone?: string;
+  regions?: string[];
+}
+
+export interface RegionMap {
+  [region: string]: CloudFunctionTrigger[];
+}
+
+export interface RegionalDeployment {
+  region: string;
+  sourceToken?: string;
+  firstFunctionDeployment?: CloudFunctionTrigger;
+  functionsToCreate: CloudFunctionTrigger[];
+  functionsToUpdate: CloudFunctionTrigger[];
+  schedulesToCreateOrUpdate: CloudFunctionTrigger[];
+}
+
+export interface DeploymentPlan {
+  regionalDeployments: RegionalDeployment[];
+  functionsToDelete: string[];
+  schedulesToDelete: string[];
+}
+
+/**
+ * Creates a map of regions to all the CloudFunctions being deployed
+ * to that region.
+ * @param projectId The project in use.
+ * @param parsedTriggers A list of all CloudFunctions in the deployment.
+ */
+export function createFunctionsByRegionMap(
```

Comment:
nit: `createFunctionsByRegionMap` sounds like you're creating functions and passing a region map...how about `groupTriggersByRegion`?

(sorry for the naming nits I think there's a lot of subtlety to these)

Reply:
I like that - and don't worry about the naming nits, its actually much appreciated. I've always found that this code path is difficult to choose good names for, partly because of semantic satiation of the word function, and partly because of the historical lack of clear typing making everything cloudy.